File: backend/tracking/focus.py
"""Focus time: how long the user actually sat down and worked.

The session itself runs client-side (src/hooks/useFocusSession.ts keeps the
timer in localStorage, as focus.js did before the port — same key, same shape).
What is tracked here is each day's total, one row per user per
day in focus.sql — so the calendar's Weekly Focus Time panel, the growth chart
and focus-type goals all read one number that survives a cleared browser.

Also here: the one-line "Focus" note attached to a calendar day, which the
Week, Day and Month views all show, so an edit in one view lands everywhere.
"""
from datetime import datetime

from backend.database import connection as db
from backend.tracking.auth import find_user


def _rows_for(username):
    return db.rows_for('focus_days', username, order='date')


def _seconds(row):
    try:
        return max(0.0, float(row.get('seconds', 0) or 0))
    except (TypeError, ValueError):
        return 0.0


def _goal_hours(row):
    try:
        return max(0.0, float(row.get('goal_hours', 0) or 0))
    except (TypeError, ValueError):
        return 0.0


def history_for(username):
    """{iso_date: {seconds, goal_hours}} for one account."""
    return {r['date']: {'seconds': _seconds(r), 'goal_hours': _goal_hours(r)}
            for r in _rows_for(username) if r.get('date')}


def record_day(username, day, seconds, goal_hours):
    """Store one day's focus total. Returns the stored record, or None.

    Never lets a stale client (an old tab with cleared localStorage, say)
    shrink a day's already-recorded total.
    """
    if not find_user(db.users(), username=username):
        return None

    # One day's row, found by its own primary key (user_id, date) rather than
    # by reading the whole ledger — every account's, every day's — and writing
    # all of it back to change one number.
    row = db.find_row('focus_days', day, user_id=username, key='date')
    kept = round(max(seconds, _seconds(row or {})), 1)

    if row is None:
        db.insert_row('focus_days', {'user_id': username, 'date': day,
                                     'seconds': kept, 'goal_hours': goal_hours})
    else:
        db.update_row('focus_days', day, {'seconds': kept, 'goal_hours': goal_hours},
                      user_id=username, key='date')
    return {'seconds': kept, 'goal_hours': goal_hours}


def history_range(username, start='', end=''):
    """{iso: {seconds, goal_hours}} within a date range, or None if no user.

    Days with no record are simply absent — nothing was tracked and nothing was
    planned.
    """
    if not find_user(db.users(), username=username):
        return None
    return {day: rec for day, rec in history_for(username).items()
            if (not start or day >= start) and (not end or day <= end)}


def total_seconds(username):
    """An account's all-time tracked focus seconds."""
    return sum(_seconds(r) for r in _rows_for(username))


def seconds_in_window(username, lo_days, hi_days, today):
    """Focused seconds recorded between `lo_days` and `hi_days` ago."""
    total = 0.0
    for row in _rows_for(username):
        try:
            d = datetime.strptime(str(row.get('date'))[:10], '%Y-%m-%d').date()
        except (ValueError, TypeError):
            continue
        if lo_days <= (today - d).days <= hi_days:
            total += _seconds(row)
    return total


# --------------------------------------------------------------------------
# The per-day focus note
# --------------------------------------------------------------------------
def day_notes(username):
    """Every saved day-focus note for a user, keyed by ISO date. None if no user."""
    if not find_user(db.users(), username=username):
        return None
    return {r['date']: r.get('text', '') for r in db.day_focus_notes()
            if r.get('user_id') == username and r.get('date')}


def set_day_note(username, day, text):
    """Upsert one day's focus text; empty text deletes the entry."""
    if not find_user(db.users(), username=username):
        return False

    db.delete_row('day_focus_notes', day, user_id=username, key='date')
    if text:
        db.insert_row('day_focus_notes',
                      {'user_id': username, 'date': day, 'text': text})
    return True

File: frontend/src/services/focus.ts
/**
 * Focus time — how long the user actually sat down and worked.
 *
 * The timer itself runs client-side; what the server keeps is each day's
 * total, so the calendar's Weekly Focus Time panel, the growth chart and
 * focus-type goals all read one number that survives a cleared browser.
 *
 * `syncDay` never lowers a day's recorded total — the backend takes the larger
 * of what it has and what it is sent. That is deliberate: an old tab whose
 * localStorage was cleared would otherwise report a smaller number and erase
 * real work. So calling it too often is harmless, and calling it with a stale
 * value is harmless too.
 *
 * Also here: the one-line note attached to a calendar day, which the Week, Day
 * and Month views all show — an edit in one view lands everywhere.
 *
 * Backend: backend/api/focus.py, backend/api/calendar.py.
 */
import { get, post } from './api';
import {
  MAX_FOCUS_SECONDS,
  MAX_GOAL_HOURS,
  MAX_LOG_MINUTES,
  MIN_GOAL_HOURS,
} from './constants';
import type { ApiResult, DayFocusNotes, FocusDay, FocusHistory } from '@/types';

/**
 * Mirror one day's focus total to the server.
 *
 * Values are clamped here as well as on the backend, so an obviously wrong
 * number is corrected before it makes a round trip rather than after.
 */
export function syncDay(
  date: string,
  focusedSeconds: number,
  goalHours: number,
): Promise<ApiResult<{ focus: FocusDay }>> {
  return post<{ focus: FocusDay }>('/api/focus_sync', {
    date,
    focused_seconds: Math.max(0, Math.min(MAX_FOCUS_SECONDS, focusedSeconds)),
    goal_hours: Math.max(MIN_GOAL_HOURS, Math.min(MAX_GOAL_HOURS, goalHours)),
  });
}

/**
 * Add work that was done and never tracked to one past day.
 *
 * The other write is `syncDay`, which is the timer mirroring itself and takes
 * the larger of the two values. This one **adds**, because it is a person
 * saying what they did rather than a client saying what it holds — a day with
 * twenty tracked minutes and two typed hours is two hours twenty. See
 * `log_day` in backend/tracking/focus.py.
 *
 * `goalHours` is only used on a day that has no record yet; a day that already
 * had a goal keeps it.
 */
export function logDay(
  date: string,
  minutes: number,
  goalHours: number,
): Promise<ApiResult<{ focus: FocusDay }>> {
  return post<{ focus: FocusDay }>('/api/focus_log', {
    date,
    minutes: Math.max(0, Math.min(MAX_LOG_MINUTES, minutes)),
    goal_hours: Math.max(MIN_GOAL_HOURS, Math.min(MAX_GOAL_HOURS, goalHours)),
  });
}

/**
 * Tracked focus for a date range, keyed by ISO date.
 *
 * Days with no record are simply absent — nothing was tracked and nothing was
 * planned — so read them with a fallback rather than expecting a full range.
 */
export function history(
  start?: string,
  end?: string,
): Promise<ApiResult<{ days: FocusHistory }>> {
  return get<{ days: FocusHistory }>('/api/focus_history', {
    start,
    end,
  });
}

// --------------------------------------------------------------------------
// The per-day focus note
// --------------------------------------------------------------------------
export function dayNotes(
): Promise<ApiResult<{ day_focus: DayFocusNotes }>> {
  return get<{ day_focus: DayFocusNotes }>('/api/day_focus');
}

/** Upsert one day's note. Empty text deletes it. Truncated to 200 chars. */
export function setDayNote(
  date: string,
  text: string,
): Promise<ApiResult<{ date: string; text: string }>> {
  return post<{ date: string; text: string }>('/api/day_focus', {
    date,
    text: text.trim().slice(0, 200),
  });
}
